source/agora/consensus/Ledger.d

```diff
@@ -677,6 +677,18 @@ public class Ledger
         }
     }
 
+    ///
+    public Amount hasPenaltyDeposit (Hash utxo) @safe nothrow
+    {
+        EnrollmentState last_enrollment;
+        if (this.enroll_man.getEnrollmentFinder()(utxo, last_enrollment) && last_enrollment.slashed_height != 0)
+            return 0.coins;
+        UTXO utxo_val;
+        if (!this.peekUTXO(utxo, utxo_val))
+            return 0.coins;
+        return this.params.SlashPenaltyAmount;
```

Comment:
TX validation ensures that freezing fee was paid for each frozen output. 

Freezing fee will still be controlled by the network only if the frozen output is still unspent and validator was never slashed while enrolled with that UTXO. Once a validator is slashed, freezing fee is paid to the commons budget as penalty in the next payout block.

If validator finishes the cycle without getting slashed it is allowed to spend the frozen utxo along with the freezing fee it paid in the freezing TX.